Project does not compile on Linux + clang/gcc

**OS:** Arch Linux, Kernel 4.20
**Compiler:** clang-7, gcc 8.2.1

**Problem:** The project cannot be compiled due to instantiations of abstract classes
**How to reproduce:** The only thing I did was to call `cmake` and run `make`, no modification or anything

**Details:** Many, many compiler errors like these: (over 60)

```
/usr/include/c++/8.2.1/bits/unique_ptr.h:831:30: error: invalid new-expression of abstract class type ‘TextMsg’
     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }

/home/kitty/git/halley/src/tests/network/src/test_stage.cpp:39:7: error: ‘void TextMsg::deserializeFrom(gsl::span<const gsl::byte>)’ marked ‘override’, but does not override
  void deserializeFrom(gsl::span<const gsl::byte> src) override
       ^~~~~~~~~~~~~~~
```

--- a/src/tests/audio/assets_src/audio_event/a1.yaml
+++ b/src/tests/audio/assets_src/audio_event/a1.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ a1.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/a1s.yaml
+++ b/src/tests/audio/assets_src/audio_event/a1s.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ a1s.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/b1.yaml
+++ b/src/tests/audio/assets_src/audio_event/b1.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ b1.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/c1.yaml
+++ b/src/tests/audio/assets_src/audio_event/c1.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ c1.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/c1s.yaml
+++ b/src/tests/audio/assets_src/audio_event/c1s.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ c1s.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/d1.yaml
+++ b/src/tests/audio/assets_src/audio_event/d1.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ d1.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/d1s.yaml
+++ b/src/tests/audio/assets_src/audio_event/d1s.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ d1s.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/e1.yaml
+++ b/src/tests/audio/assets_src/audio_event/e1.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ e1.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/f1.yaml
+++ b/src/tests/audio/assets_src/audio_event/f1.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ f1.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/f1s.yaml
+++ b/src/tests/audio/assets_src/audio_event/f1s.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ f1s.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/g1.yaml
+++ b/src/tests/audio/assets_src/audio_event/g1.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ g1.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/g1s.yaml
+++ b/src/tests/audio/assets_src/audio_event/g1s.yaml
@@ -1,0 +1,4 @@
+actions:
+  - type: play
+    clips: [ g1s.ogg ]
+    group: key

--- a/src/tests/audio/assets_src/audio_event/loveshadow.yaml
+++ b/src/tests/audio/assets_src/audio_event/loveshadow.yaml
@@ -1,0 +1,5 @@
+actions:
+  - type: play
+    clips: [ Loveshadow_-_Marcos_Theme.ogg ]
+    group: music
+    loop: true

--- a/src/tests/audio/src/main.cpp
+++ b/src/tests/audio/src/main.cpp
@@ -4,7 +4,7 @@
 using namespace Halley;
 
 void initOpenGLPlugin(IPluginRegistry &registry);
-void initSDLSystemPlugin(IPluginRegistry &registry);
+void initSDLSystemPlugin(IPluginRegistry &registry, Maybe<String> cryptKey);
 void initSDLAudioPlugin(IPluginRegistry &registry);
 void initSDLInputPlugin(IPluginRegistry &registry);
 
@@ -20,16 +20,16 @@ class AudioTestGame final : public Game
 public:
 	int initPlugins(IPluginRegistry &registry) override
 	{
-		initSDLSystemPlugin(registry);
+		initSDLSystemPlugin(registry, {});
 		initSDLAudioPlugin(registry);
 		initSDLInputPlugin(registry);
 		initOpenGLPlugin(registry);
 		return HalleyAPIFlags::Video | HalleyAPIFlags::Audio | HalleyAPIFlags::Input;
 	}
 
-	void initResourceLocator(Path dataPath, ResourceLocator& locator) override
+	void initResourceLocator(const Path& gamePath, const Path& assetsPath, const Path& unpackedAssetsPath, ResourceLocator& locator) override
 	{
-		locator.addFileSystem(dataPath);
+		locator.addFileSystem(unpackedAssetsPath);
 	}
 
 	std::unique_ptr<Stage> makeStage(StageID id) override
@@ -52,7 +52,7 @@ public:
 		return "halley/audio-test";
 	}
 
-	bool isDevBuild() const override
+	bool isDevMode() const override
 	{
 		return true;
 	}
@@ -60,7 +60,7 @@ public:
 	std::unique_ptr<Stage> startGame(const HalleyAPI* api) override
 	{
 		api->audio->startPlayback();
-		api->video->setWindow(WindowDefinition(WindowType::Window, Vector2i(1280, 720), getName()), true);
+		api->video->setWindow(WindowDefinition(WindowType::Window, Vector2i(1280, 720), getName()));
 		return std::make_unique<TestStage>();
 	}
 };

--- a/src/tests/audio/src/test_stage.cpp
+++ b/src/tests/audio/src/test_stage.cpp
@@ -5,7 +5,7 @@ using namespace Halley;
 
 void TestStage::init()
 {
-	getAudioAPI().playMusic(getResource<AudioClip>("Loveshadow_-_Marcos_Theme.ogg"));
+	getAudioAPI().play(getResource<AudioClip>("Loveshadow_-_Marcos_Theme.ogg"), AudioPosition());
 }
 
 void TestStage::onVariableUpdate(Time time)
@@ -32,7 +32,7 @@ void TestStage::onVariableUpdate(Time time)
 	auto noteSamples = std::array<String, 12>{{ "c1", "c1s", "d1", "d1s", "e1", "f1", "f1s", "g1", "g1s", "a1", "a1s", "b1" }};
 	for (int i = 0; i < 12; ++i) {
 		if (key->isButtonPressed(noteKeys[i])) {
-			getAudioAPI().playUI(getResources().get<AudioClip>(noteSamples[i] + ".ogg"), 1.0f, pan);
+			getAudioAPI().play(getResources().get<AudioClip>(noteSamples[i] + ".ogg"), AudioPosition::makeUI(pan));
 		}
 	}
 }

--- a/src/tests/audio/src/test_stage.cpp
+++ b/src/tests/audio/src/test_stage.cpp
@@ -5,7 +5,7 @@ using namespace Halley;
 
 void TestStage::init()
 {
-	getAudioAPI().play(getResource<AudioClip>("Loveshadow_-_Marcos_Theme.ogg"), AudioPosition());
+	getAudioAPI().playMusic("loveshadow");
 }
 
 void TestStage::onVariableUpdate(Time time)
@@ -32,7 +32,7 @@ void TestStage::onVariableUpdate(Time time)
 	auto noteSamples = std::array<String, 12>{{ "c1", "c1s", "d1", "d1s", "e1", "f1", "f1s", "g1", "g1s", "a1", "a1s", "b1" }};
 	for (int i = 0; i < 12; ++i) {
 		if (key->isButtonPressed(noteKeys[i])) {
-			getAudioAPI().play(getResources().get<AudioClip>(noteSamples[i] + ".ogg"), AudioPosition::makeUI(pan));
+			getAudioAPI().postEvent(noteSamples[i], AudioPosition::makeUI(pan));
 		}
 	}
 }

--- a/src/tests/entity/src/main.cpp
+++ b/src/tests/entity/src/main.cpp
@@ -4,7 +4,7 @@
 using namespace Halley;
 
 void initOpenGLPlugin(IPluginRegistry &registry);
-void initSDLSystemPlugin(IPluginRegistry &registry);
+void initSDLSystemPlugin(IPluginRegistry &registry, Maybe<String> cryptKey);
 void initSDLAudioPlugin(IPluginRegistry &registry);
 void initSDLInputPlugin(IPluginRegistry &registry);
 
@@ -33,16 +33,16 @@ class EntityTestGame final : public Game
 public:
 	int initPlugins(IPluginRegistry &registry) override
 	{
-		initSDLSystemPlugin(registry);
+		initSDLSystemPlugin(registry, {});
 		initSDLAudioPlugin(registry);
 		initSDLInputPlugin(registry);
 		initOpenGLPlugin(registry);
 		return HalleyAPIFlags::Video | HalleyAPIFlags::Audio | HalleyAPIFlags::Input;
 	}
 
-	void initResourceLocator(Path dataPath, ResourceLocator& locator) override
+	void initResourceLocator(const Path& gamePath, const Path& assetsPath, const Path& unpackedAssetsPath, ResourceLocator& locator) override
 	{
-		locator.addFileSystem(dataPath);
+		locator.addFileSystem(unpackedAssetsPath);
 	}
 
 	std::unique_ptr<Stage> makeStage(StageID id) override
@@ -69,14 +69,14 @@ public:
 		return "halley/entity-test";
 	}
 
-	bool isDevBuild() const override
+	bool isDevMode() const override
 	{
 		return true;
 	}
 
 	std::unique_ptr<Stage> startGame(const HalleyAPI* api) override
 	{
-		api->video->setWindow(WindowDefinition(WindowType::Window, Vector2i(1280, 720), getName()), true);
+		api->video->setWindow(WindowDefinition(WindowType::Window, Vector2i(1280, 720), getName()));
 		return std::make_unique<TestStage>();
 	}
 };

--- a/src/tests/entity/src/systems/render_system.cpp
+++ b/src/tests/entity/src/systems/render_system.cpp
@@ -12,13 +12,13 @@ public:
 		for (auto& e : mainFamily) {
 			auto& sprite = e.sprite.sprite;
 			sprite.setPos(e.position.position);
-			spritePainter.add(sprite, 0, e.sprite.layer, sprite.getPosition().y);
+			spritePainter.add(sprite, 1, e.sprite.layer, sprite.getPosition().y);
 		}
 
 		rc.bind([&] (Painter& painter)
 		{
 			painter.clear(Colour());
-			spritePainter.draw(0, painter);
+			spritePainter.draw(1, painter);
 			painter.flush();
 		});
 	}

--- a/src/tests/entity/src/test_stage.cpp
+++ b/src/tests/entity/src/test_stage.cpp
@@ -6,7 +6,10 @@ using namespace Halley;
 void TestStage::init()
 {
 	world = createWorld("sample_test_world", createSystem);
-	statsView = std::make_unique<WorldStatsView>(*getAPI().core, *world);
+	statsView = std::make_unique<WorldStatsView>(*getAPI().core);
+	auto stolen = world.release();
+	world.reset(stolen);
+	statsView->setWorld(stolen);
 }
 
 void TestStage::onFixedUpdate(Time time)

--- a/src/tests/entity/src/test_stage.cpp
+++ b/src/tests/entity/src/test_stage.cpp
@@ -7,9 +7,7 @@ void TestStage::init()
 {
 	world = createWorld("sample_test_world", createSystem);
 	statsView = std::make_unique<WorldStatsView>(*getAPI().core);
-	auto stolen = world.release();
-	world.reset(stolen);
-	statsView->setWorld(stolen);
+	statsView->setWorld(world.get());
 }
 
 void TestStage::onFixedUpdate(Time time)

--- a/src/tests/network/src/main.cpp
+++ b/src/tests/network/src/main.cpp
@@ -14,9 +14,9 @@ public:
 		return HalleyAPIFlags::Video | HalleyAPIFlags::Audio | HalleyAPIFlags::Input;
 	}
 
-	void initResourceLocator(Path dataPath, ResourceLocator& locator) override
+	void initResourceLocator(const Path& gamePath, const Path& assetsPath, const Path& unpackedAssetsPath, ResourceLocator& locator) override
 	{
-		locator.addFileSystem(dataPath);
+		locator.addFileSystem(unpackedAssetsPath);
 	}
 
 	String getName() const override
@@ -29,14 +29,14 @@ public:
 		return "halley/network-test";
 	}
 
-	bool isDevBuild() const override
+	bool isDevMode() const override
 	{
 		return true;
 	}
 
 	std::unique_ptr<Stage> startGame(const HalleyAPI* api) override
 	{
-		api->video->setWindow(WindowDefinition(WindowType::Window, Vector2i(1280, 720), getName()), false);
+		api->video->setWindow(WindowDefinition(WindowType::Window, Vector2i(1280, 720), getName()));
 		return std::make_unique<TestStage>();
 	}
 };

--- a/src/tests/network/src/test_stage.cpp
+++ b/src/tests/network/src/test_stage.cpp
@@ -2,6 +2,7 @@
 #include "halley/text/string_converter.h"
 #include "halley/net/connection/iconnection.h"
 #include "halley/net/connection/network_message.h"
+#include "halley/net/connection/message_queue_tcp.h"
 
 using namespace Halley;
 
@@ -9,9 +10,9 @@ using namespace Halley;
 class NoOpMsg : public NetworkMessage
 {
 public:
-	size_t getSerializedSize() const override { return 0; }	
-	void serializeTo(gsl::span<gsl::byte> /*dst*/) const override {}	
-	void deserializeFrom(gsl::span<const gsl::byte> /*src*/) override {}
+	NoOpMsg(gsl::span<const gsl::byte> src) {}
+
+	void serialize(Serializer&) const override {}
 };
 
 class TextMsg : public NetworkMessage
@@ -24,21 +25,16 @@ public:
 		: str(str)
 	{}
 
+	TextMsg(gsl::span<const gsl::byte> src) {
+		Deserializer d(src);
+		d >> str;
+	}
+
 	String getString() const { return str; }
 
-	size_t getSerializedSize() const override
+	void serialize(Serializer& s) const override
 	{
-		return str.size();
-	}
-	
-	void serializeTo(gsl::span<gsl::byte> dst) const override
-	{
-		memcpy(dst.data(), str.c_str(), str.size());
-	}
-	
-	void deserializeFrom(gsl::span<const gsl::byte> src) override
-	{
-		str = String(reinterpret_cast<const char*>(src.data()), src.size());
+		s << str;
 	}
 
 private:
@@ -81,14 +77,14 @@ void TestStage::updateNetwork()
 		if (key->isButtonPressed(Keys::S)) {
 			// Server
 			isClient = false;
-			network = getNetworkAPI().createService(4113);
+			network = getNetworkAPI().createService(NetworkProtocol::TCP, 4113);
 			network->setAcceptingConnections(true);
 			std::cout << "Listening..." << std::endl;
 		}
 		else if (key->isButtonPressed(Keys::C)) {
 			// Client
 			isClient = true;
-			network = getNetworkAPI().createService(0);
+			network = getNetworkAPI().createService(NetworkProtocol::TCP, 0);
 			setConnection(network->connect("127.0.0.1", 4113));			
 			std::cout << "Connecting as client." << std::endl;
 		}
@@ -109,7 +105,7 @@ void TestStage::updateNetwork()
 				}
 			}
 			
-			if (connection->getStatus() == ConnectionStatus::Open) {
+			if (connection->getStatus() == ConnectionStatus::Connected) {
 				if (key->isButtonPressed(Keys::Space)) {
 					msgs->enqueue(std::make_unique<TextMsg>("ding"), 1);
 				}
@@ -152,7 +148,7 @@ void TestStage::setConnection(std::shared_ptr<Halley::IConnection> conn)
 	auto base = unstable ? std::make_shared<InstabilitySimulator>(conn, 0.1f, 0.03f, 0.1f, 0.05f) : conn;
 	connection = std::make_shared<ReliableConnection>(base);
 	
-	msgs = std::make_unique<MessageQueue>(connection);
+	msgs = std::make_unique<MessageQueueTCP>(connection);
 	msgs->setChannel(0, ChannelSettings(false, false, false));
 	msgs->setChannel(1, ChannelSettings(true, false, false));
 	msgs->setChannel(2, ChannelSettings(false, true, false));

